### PR TITLE
fix(query): fix max_files bug when force=false

### LINKED ISSUE #10800: fix purge=false err
expect 6

_Originally posted by @youngsofun in https://github.com/datafuselabs/databend/pull/10790#discussion_r1149994943_

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

purge = false force = false: 

It does not change after copying all files

purge = false force = true: 

 only copy the max_files nums file.

Closes #10800

### CODE REVIEW

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh [hunk 1]
```diff
@@ -41,6 +41,14 @@ EOF
 echo "copy into test_max_files_force_true from 'fs:///tmp/05_02_06/' FILE_FORMAT = (type = CSV) max_files=2 force=true" | $MYSQL_CLIENT_CONNECT
 echo "select count(*) from test_max_files_force_true" | $MYSQL_CLIENT_CONNECT
 
+echo "copy into test_max_files_force_true from 'fs:///tmp/05_02_06/' FILE_FORMAT = (type = CSV) max_files=2 force=true" | $MYSQL_CLIENT_CONNECT
+echo "select count(*) from test_max_files_force_true" | $MYSQL_CLIENT_CONNECT
+
+echo "copy into test_max_files_force_true from 'fs:///tmp/05_02_06/' FILE_FORMAT = (type = CSV) max_files=2 force=true" | $MYSQL_CLIENT_CONNECT
+echo "select count(*) from test_max_files_force_true" | $MYSQL_CLIENT_CONNECT
```
Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh:48]: should return 4 not  6?

File: tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh [hunk 2]
```diff
@@ -41,6 +41,14 @@ EOF
...
+
+echo "copy into test_max_files_force_false from 'fs:///tmp/05_02_06/' FILE_FORMAT = (type = CSV) max_files=2 force=false" | $MYSQL_CLIENT_CONNECT
```
Comment (youngsofun) [inline tests/suites/1_stateful/05_formats/05_02_csv/05_02_06_csv_max_files.sh:50]: should return 4? not 6?

File: src/query/catalog/src/table_context.rs [hunk 3]
```diff
@@ -141,7 +141,7 @@ pub trait TableContext: Send + Sync {
         catalog_name: &str,
         database_name: &str,
         table_name: &str,
-        files: Vec<StageFileInfo>,
+        mut files: Vec<StageFileInfo>,
```
Comment (BohuTANG) [inline src/query/catalog/src/table_context.rs:144]:
```suggestion
        files: Vec<StageFileInfo>,
```
please don't make it to mut.

File: src/query/service/src/sessions/query_ctx.rs [hunk 4]
```diff
@@ -509,45 +510,58 @@ impl TableContext for QueryContext {
...
+                    let file = StageFileInfo {
+                        path: file.path.clone(),
+                        size: file.size.clone(),
+                        md5: file.md5.clone(),
+                        last_modified: file.last_modified.clone(),
+                        etag: file.etag.clone(),
+                        status: StageFileStatus::NeedCopy,
+                        creator: file.creator.clone(),
+                    };
+                    results.push(file);
```
Comment (youngsofun) [inline src/query/service/src/sessions/query_ctx.rs:562]: may push more than max_files?

File: src/query/service/src/sessions/query_ctx.rs [hunk 5]
```diff
@@ -498,7 +499,7 @@ impl TableContext for QueryContext {
         catalog_name: &str,
         database_name: &str,
         table_name: &str,
-        files: Vec<StageFileInfo>,
+        mut files: Vec<StageFileInfo>,
```
Comment (BohuTANG) [inline src/query/service/src/sessions/query_ctx.rs:502]: We can make a copy, but please don't make the argument to `mut`. This is bit tricky.

File: src/query/service/src/sessions/query_ctx.rs [hunk 6]
```diff
@@ -509,45 +510,58 @@ impl TableContext for QueryContext {
...
 
         let mut limit: usize = 0;
+        let max_files = max_files.unwrap_or(MAX_QUERY_COPIED_FILES_NUM);
+        let max_copied_files = min(MAX_QUERY_COPIED_FILES_NUM, max_files);
         let mut copied_files = BTreeMap::new();
-        for chunk in files.chunks(MAX_QUERY_COPIED_FILES_NUM) {
+
+        let mut results = Vec::with_capacity(files.len());
+
+        for chunk in files.chunks_mut(max_copied_files) {
```
Comment (BohuTANG) [inline src/query/service/src/sessions/query_ctx.rs:519]: Why need `chunks_mut`?

File: src/query/service/src/sessions/query_ctx.rs [hunk 7]
```diff
@@ -502,52 +503,66 @@ impl TableContext for QueryContext {
...
         let tenant = self.get_tenant();
+        let mut files = files.clone();
         let catalog = self.get_catalog(catalog_name)?;
         let table = catalog
             .get_table(&tenant, database_name, table_name)
             .await?;
         let table_id = table.get_id();
 
         let mut limit: usize = 0;
+        let max_files = max_files.unwrap_or(MAX_QUERY_COPIED_FILES_NUM);
```
Comment (youngsofun) [inline src/query/service/src/sessions/query_ctx.rs:514]: https://github.com/datafuselabs/databend/pull/10790/files#diff-7c72c74465abbbfa3b0827662f229e83fa78c5473ad04f1cd4ca444d4088a47bR565

it is called with max_files=None?

File: src/query/service/src/sessions/query_ctx.rs [hunk 8]
```diff
@@ -502,52 +503,66 @@ impl TableContext for QueryContext {
...
-                    None => {
-                        // etag is none, compare with content_length and last_modified.
-                        if copied_file.content_length == file.size
-                            && copied_file.last_modified == Some(file.last_modified)
-                        {
-                            file.status = StageFileStatus::AlreadyCopied;
-                        }
+                }
+                if file.status == StageFileStatus::NeedCopy {
+                    if limit == max_files {
```
Comment (youngsofun) [inline src/query/service/src/sessions/query_ctx.rs:550]: push -> +1 -> check is better

File: src/query/service/src/sessions/query_ctx.rs (hunk 7 above)
Comment (youngsofun) [reply src/query/service/src/sessions/query_ctx.rs:514]: https://github.com/datafuselabs/databend/issues/10803 we can do it later

File: src/query/service/src/sessions/query_ctx.rs (hunk 6 above)
Comment (youngsofun) [reply src/query/service/src/sessions/query_ctx.rs:520]: seems "all files" is only used in logs

**Review by youngsofun — APPROVED** [review]

**Pull request merged**